File: server/routes/skills.js
import { readFile, writeFile, readdir } from 'node:fs/promises';
import { join } from 'node:path';
import { jsonResponse, parseJsonBody, ensureDir, fileExists, requireFields, readJsonOrNull, safePath } from '../middleware.js';

// --- Helpers (only used by skills routes) ---

function simpleHash(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = ((hash << 5) - hash) + str.charCodeAt(i);
    hash |= 0;
  }
  return hash.toString(36);
}

function parseDescription(text) {
  const match = text.match(/^---\s*\n([\s\S]*?)\n---/);
  if (!match) return '';
  const fm = match[1];
  // Try quoted description first
  let descMatch = fm.match(/description:\s*"([^"]*(?:"[^"]*?)*)"/m) ||
                  fm.match(/description:\s*'([^']*)'/m);
  if (!descMatch) {
    // Unquoted single-line
    descMatch = fm.match(/description:\s*(.+)$/m);
  }
  if (!descMatch) {
    // Fallback: first non-heading line after frontmatter
    const firstLine = text.split('\n---')[1]?.split('\n').find(l => l.trim() && !l.startsWith('#'));
    if (firstLine) return firstLine.trim().slice(0, 120);
    return '';
  }
  let desc = descMatch[1].trim();
  desc = desc.replace(/\s*TRIGGER\s+on:.*$/i, '').replace(/\s*Use when:.*$/i, '');
  let end = desc.indexOf('. ');
  if (end > 0) {
    const second = desc.indexOf('. ', end + 2);
    if (second > 0 && second < 200) end = second + 1;
    else end = end + 1;
  }
  if (end > 0 && end < 200) desc = desc.slice(0, end);
  else if (desc.length > 200) desc = desc.slice(0, 200) + '...';
  return desc;
}

export async function handleSkills(method, pathname, req, res, url, ctx) {
  const { projectPath } = ctx;

  // GET /api/skills
  if (method === 'GET' && pathname === '/api/skills') {
    const results = [];

    // Discover skills from .claude/skills/
    const skillsDir = join(projectPath, '.claude', 'skills');
    try {
      const dirs = await readdir(skillsDir, { withFileTypes: true });
      for (const entry of dirs) {
        if (!entry.isDirectory() || entry.name === 'orchestrator') continue;
        let description = '';
        // Try SKILL.md then skill.md
        for (const filename of ['SKILL.md', 'skill.md']) {
          const filePath = join(skillsDir, entry.name, filename);
          if (await fileExists(filePath)) {
            try {
              const text = await readFile(filePath, 'utf-8');
              description = parseDescription(text);
            } catch { /* ignore */ }
            break;
          }
        }
        results.push({ name: entry.name, description, type: 'skill' });
      }
    } catch (err) {
      if (err.code !== 'ENOENT') console.error('Skills scan error:', err.message);
    }

    // Discover agents from .claude/agents/
    const agentsDir = join(projectPath, '.claude', 'agents');
    try {
      const files = await readdir(agentsDir, { withFileTypes: true });
      for (const entry of files) {
        if (!entry.isFile() || !entry.name.endsWith('.md')) continue;
        const agentName = entry.name.replace(/\.md$/, '');
        let description = '';
        try {
          const text = await readFile(join(agentsDir, entry.name), 'utf-8');
          description = parseDescription(text);
        } catch { /* ignore */ }
        results.push({ name: agentName, description, type: 'agent' });
      }
    } catch (err) {
      if (err.code !== 'ENOENT') console.error('Agents scan error:', err.message);
    }

    results.sort((a, b) => a.name.localeCompare(b.name));
    jsonResponse(res, 200, results);
    return true;
  }

  // POST /api/skills/deploy
  if (method === 'POST' && pathname === '/api/skills/deploy') {
    const body = await parseJsonBody(req);
    const { skillName, filename, content } = body;
    const err = requireFields(body, 'skillName', 'filename', 'content');
    if (err) { jsonResponse(res, 400, { error: err }); return true; }
    const skillDir = safePath(projectPath, '.claude', 'skills', skillName);
    if (!skillDir) { jsonResponse(res, 400, { error: 'Invalid skill name' }); return true; }
    await ensureDir(skillDir);

    // Check hash to skip if unchanged
    const hash = simpleHash(content);
    const hashPath = safePath(skillDir, '.hash');
    if (!hashPath) { jsonResponse(res, 400, { error: 'Invalid path' }); return true; }
    try {
      const existingHash = await readFile(hashPath, 'utf-8');
      if (existingHash.trim() === hash) {
        jsonResponse(res, 200, { ok: true, deployed: false });
        return true;
      }
    } catch { /* no hash file yet */ }

    const skillFilePath = safePath(skillDir, filename);
    if (!skillFilePath) { jsonResponse(res, 400, { error: 'Invalid filename' }); return true; }
    await writeFile(skillFilePath, content, 'utf-8');
    await writeFile(hashPath, hash, 'utf-8');
    jsonResponse(res, 200, { ok: true, deployed: true });
    return true;
  }

  // POST /api/agents/deploy
  if (method === 'POST' && pathname === '/api/agents/deploy') {
    const body = await parseJsonBody(req);
    const { agentName, filename, content } = body;
    const err = requireFields(body, 'agentName', 'filename', 'content');
    if (err) { jsonResponse(res, 400, { error: err }); return true; }
    const agentsDir = join(projectPath, '.claude', 'agents');
    await ensureDir(agentsDir);

    const agentDir = safePath(agentsDir, agentName);
    if (!agentDir) { jsonResponse(res, 400, { error: 'Invalid agent name' }); return true; }
    await ensureDir(agentDir);
    const agentFilePath = safePath(agentDir, filename);
    if (!agentFilePath) { jsonResponse(res, 400, { error: 'Invalid filename' }); return true; }
    await writeFile(agentFilePath, content, 'utf-8');
    jsonResponse(res, 200, { ok: true, deployed: true });
    return true;
  }

  // GET /api/skills-config
  if (method === 'GET' && pathname === '/api/skills-config') {
    const configPath = join(projectPath, '.maestro', 'skills.json');
    const result = await readJsonOrNull(configPath);
    jsonResponse(res, 200, result ? result.data : {});
    return true;
  }

  // PUT /api/skills-config
  if (method === 'PUT' && pathname === '/api/skills-config') {
    const body = await parseJsonBody(req);
    const configDir = join(projectPath, '.maestro');
    await ensureDir(configDir);
    await writeFile(join(configDir, 'skills.json'), JSON.stringify(body, null, 2), 'utf-8');
    jsonResponse(res, 200, { ok: true });
    return true;
  }

  return false;
}

File: apps/tauri/src-tauri/src/commands/skills.rs
use serde_json::Value;
use std::fs;
use std::path::PathBuf;
use tauri::State;

use crate::ProjectPath;

#[derive(serde::Serialize)]
pub struct SkillInfo {
    pub name: String,
    pub description: String,
    #[serde(rename = "type")]
    pub skill_type: String,
}

fn skills_dir(project: &str) -> PathBuf {
    PathBuf::from(project).join(".claude").join("skills")
}

fn agents_dir(project: &str) -> PathBuf {
    PathBuf::from(project).join(".claude").join("agents")
}

fn skills_config_path(project: &str) -> PathBuf {
    PathBuf::from(project)
        .join(".maestro")
        .join("skills.json")
}

#[tauri::command]
pub fn discover_skills(project: State<'_, ProjectPath>) -> Result<Vec<SkillInfo>, String> {
    let mut results = Vec::new();

    // Discover skills
    let sd = skills_dir(&project.get());
    if sd.exists() {
        if let Ok(entries) = fs::read_dir(&sd) {
            for entry in entries.flatten() {
                if entry.path().is_dir() {
                    let name = entry.file_name().to_string_lossy().to_string();
                    // Try to read description from SKILL.md or skill.md
                    let desc = read_skill_description(&entry.path());
                    results.push(SkillInfo {
                        name,
                        description: desc,
                        skill_type: "skill".to_string(),
                    });
                }
            }
        }
    }

    // Discover agents
    let ad = agents_dir(&project.get());
    if ad.exists() {
        if let Ok(entries) = fs::read_dir(&ad) {
            for entry in entries.flatten() {
                if entry.path().is_dir() {
                    let name = entry.file_name().to_string_lossy().to_string();
                    let desc = read_skill_description(&entry.path());
                    results.push(SkillInfo {
                        name,
                        description: desc,
                        skill_type: "agent".to_string(),
                    });
                }
            }
        }
    }

    Ok(results)
}

fn read_skill_description(dir: &std::path::Path) -> String {
    // Look for SKILL.md, skill.md, or AGENT.md
    for name in &["SKILL.md", "skill.md", "AGENT.md", "agent.md"] {
        let path = dir.join(name);
        if let Ok(content) = fs::read_to_string(&path) {
            // Extract first non-empty, non-heading line as description
            for line in content.lines() {
                let trimmed = line.trim();
                if !trimmed.is_empty() && !trimmed.starts_with('#') {
                    return trimmed.to_string();
                }
            }
        }
    }
    String::new()
}

#[tauri::command]
pub fn deploy_skill(
    skill_name: String,
    filename: String,
    content: String,
    project: State<'_, ProjectPath>,
) -> Result<bool, String> {
    let dir = skills_dir(&project.get()).join(&skill_name);
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    let path = dir.join(&filename);

    // Check if content is the same — skip write if unchanged
    if let Ok(existing) = fs::read_to_string(&path) {
        if existing == content {
            return Ok(false); // Not deployed (already up to date)
        }
    }

    fs::write(&path, &content).map_err(|e| e.to_string())?;
    Ok(true) // Deployed
}

#[tauri::command]
pub fn deploy_agent(
    agent_name: String,
    filename: String,
    content: String,
    project: State<'_, ProjectPath>,
) -> Result<bool, String> {
    let dir = agents_dir(&project.get()).join(&agent_name);
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    let path = dir.join(&filename);

    if let Ok(existing) = fs::read_to_string(&path) {
        if existing == content {
            return Ok(false);
        }
    }

    fs::write(&path, &content).map_err(|e| e.to_string())?;
    Ok(true)
}

#[tauri::command]
pub fn read_skills_config(project: State<'_, ProjectPath>) -> Result<Value, String> {
    let path = skills_config_path(&project.get());
    if !path.exists() {
        return Ok(serde_json::json!(null));
    }
    let content = fs::read_to_string(&path).map_err(|e| e.to_string())?;
    serde_json::from_str(&content).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn write_skills_config(config: Value, project: State<'_, ProjectPath>) -> Result<bool, String> {
    let path = skills_config_path(&project.get());
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    let json = serde_json::to_string_pretty(&config).map_err(|e| e.to_string())?;
    fs::write(&path, json).map_err(|e| e.to_string())?;
    Ok(true)
}
